gprofiler/merge.py

```diff
@@ -82,7 +82,85 @@ def collapse_stack(stack: str, comm: str) -> str:
     return ";".join(funcs)
 
 
-def parse_perf_script(script: str):
+def merge_global_perfs(
+    raw_fp_perf: Optional[str], raw_dwarf_perf: Optional[str]
+) -> Tuple[DefaultDict[int, MutableMapping[str, int]], Dict[int, str]]:
+    # The FP perf is used here as the "main" perf, to which the Dwarf perf is scaled.
+    merged_pid_to_stacks_counters: DefaultDict[int, MutableMapping[str, int]] = defaultdict(Counter)
+    fp_perf, fp_pid_to_name = parse_perf_script(raw_fp_perf)
+    dwarf_perf, dwarf_pid_to_name = parse_perf_script(raw_dwarf_perf)
+    dwarf_pid_to_name.update(fp_pid_to_name)
+    merged_pid_to_name = dwarf_pid_to_name
+
+    if raw_fp_perf is None:
+        return dwarf_perf, merged_pid_to_name
+    elif raw_dwarf_perf is None:
+        return fp_perf, merged_pid_to_name
+
+    total_fp_samples = sum([sum(stacks.values()) for stacks in fp_perf.values()])
+    total_dwarf_samples = sum([sum(stacks.values()) for stacks in dwarf_perf.values()])
+    fp_to_dwarf_total_sample_count_ratio = total_fp_samples / total_dwarf_samples
+
+    add_highest_avg_depth_stacks_per_process(
+        dwarf_perf, fp_perf, fp_to_dwarf_total_sample_count_ratio, merged_pid_to_stacks_counters
+    )
+    add_missing_dwarf_stacks(dwarf_perf, fp_to_dwarf_total_sample_count_ratio, merged_pid_to_stacks_counters)
+
+    return merged_pid_to_stacks_counters, merged_pid_to_name
+
+
+def add_highest_avg_depth_stacks_per_process(
+    dwarf_perf, fp_perf, fp_to_dwarf_total_sample_count_ratio, merged_pid_to_stacks_counters
+):
+    for pid, fp_collapsed_stacks_counters in fp_perf.items():
+        if pid not in dwarf_perf:
```

Comment:
What about the other way (only in `dwarf` and not in `fp`)

Reply:
I'm actually not sure this is the behavior we want. Since both `perf`s run concurrently and generate their samples at different intervals, we shouldn't inject samples from one into another.
Let's imagine a simple case - 1 core, 2 apps running at 100%. `fp` and `dwarf` were started at different intervals so there's a given offset between their interrupt (say, 30 ms). The `fp` might "fall" onto stacks from app 1, and the `dwarf` on stacks from app 2. If you merge in this case - then we're taking double the stacks for no good reason.

@DanielShaulov I recall we've already discussed it but I don't remember what was the outcome :sweat_smile:

Reply:
Whatever we pick - like suggested in https://github.com/Granulate/gprofiler/pull/77/files#r636259102, please add a `logger.debug` call giving the numbers:
* total collected fp
* total collected dwarf
* total merged

Reply:
What you are saying is that we have 2 apps running at 100%, and the `fp` graph would only get samples from one of them?
Regardless of dwarf samples, this would mean current gprofiler would show a very skewed result (all samples from app1 vs 50-50).
If you mean that it might align perfectly with context switches, then we do need to set our frequency to something not divisible by 10, it is a common suggestion (like our default of 3999 in gserver, so maybe just 9 here? Or increase to 19?).
We should be able to trust the `fp` graph to be a good representation of how much CPU each process takes.

In any case - I didn't review the code yet to see if it fits the algorithm, but what it should do is treat the `fp` run as the "master" flamegraph, and assume all process proportions in it are "correct", then everything else is scaled to it - so a per-process perf results are scaled to the amount of samples in the `fp` graph for that process.
Dwarf is no different, we should look at the amount of samples of that process in the `dwarf` and the `fp` graph and scale to fit the `fp`. If it doesn't exist in the `fp` - it should not be in the end result at all. If it exists only in the `fp`, we would just the `fp` result for it. (Exactly the same logic is true for java and python currently)

Reply:
> If you mean that it might align perfectly with context switches, then we do need to set our frequency to something not divisible by 10.

My example was "catastrophic" to demonstrate the issue at its full extent, but I assume that even less catastrophic "flows" can yield skewed results.
I don't mean the extreme case that we happen to sample at the context switch times. Changing to 11hz/9hz/19hz is a good idea nonetheless (I've got no idea why I'd set it to 10 in the first place...) - but it won't mitigate what I have in mind.

> it should not be in the end result at all.

Yes, that's what I meant in https://github.com/Granulate/gprofiler/pull/77/files#r636267222 and that's the behavior now. So all good in that regard...